### ISSUE TITLE
Learning to Pool in Graph Neural Networks for Extrapolation

### ISSUE THREAD

**Issue by hello-im-yj** [One-Day-One-Paper/One-Day-One-Paper#201]
Xu et al. (2021) studied the impact of the choice of pooling functions on the ability of a neural network to extrapolate: The findings in Xu et al. (2021) raise a natural `question`; which pooling functions should be used for a given problem in order to make GNNs constructed with them successfully extrapolate for out-of-distribution data?

`proposed method` : Generalized Norm-based Pooling (GNP), formulates the pooling functions as a generic $L^p$ norm-like function (including negative p as well), and learns the parameters inside the pooling functions in an end-to-end fashion.

→ previous learning-based pooling methods는 특정 task(혹은 분류 중심 task)만을 위해 디자인 된 반면에, 본 논문에서 제안하는 모델은 임의의 task에 적용가능하며 GNN의 추론 능력을 향상시킴.

# 3 MAIN CONTRIBUTION: GENERALIZED NORM-BASED POOLING

3.1 GENERALIZATION OF BASIC POOLING FUNCTIONS

- Lp norm
    
    > [https://arxiv.org/pdf/1311.1780.pdf](https://arxiv.org/pdf/1311.1780.pdf) : Learned-Norm Pooling for Deep Feedforward
    and Recurrent Neural Networks (Caglar Gulcehre, Kyunghyun Cho, Razvan Pascanu, and Yoshua Bengio)
    > 

$L^p$**-norm function**

![image](https://user-images.githubusercontent.com/63237947/172024906-09349d90-ca0f-4b57-ab32-633d58164f9d.png)

- $p$, $q$ : learnable parameters
    - $p ∈ \mathbb{R} /\ \{0\}$ : 0을 제외한 실수
    - $q ∈ \mathbb{R}$
- includes the basic pooling functions (`sum`, `mean`, `max`, and `min`) as special cases.
    - $GNP_j (V)$ is equivalent to elementwise `sum` when $(p, q) = (1, 0)$ and elementwise `mean` when $(p, q) = (1, 1)$.

3.2 HANDLING OF NEGATIVE p

- negative $p$ 를 다루기 위해 : node feature $V$의 dementsion을 1/2로 나누어 절반은 GNP^+(p>0) 를 절반은 GNP^-(p<0) 로 학습한 후 → 이를 single linier layer를 통해 mix한 후 final output을 얻음.

3.3 STABILIZATION OF TRAINING PROCESSES

- 안정화를 위한 Trick : 1)small tolerance term $\epsilon$ 를 도입하여 gradient explosion을 막음 2)p가 크면 수렴하지 않기 때문에 p값을 [0,50]사이로 클리핑 3)p값이 GNP에 더 큰 영향을 미치기 때문에 (p+,p-)와 (q+,q-) 학습에 다른 learning rate를 적용
